### PR TITLE
Added config option proxy.config.thread.max_heartbeat_mseconds

### PR DESCRIPTION
**Pull request by duke8253**

### CODE REVIEW

**Comment by zwoop** [comment]
Hmmm, so this changes it from 60 to 10 (default) ? I bet this will show up as more CPU used on idle boxes ? Any other issues with that?

**Comment by zwoop** [comment]
Also, remember to set Milestone. ;-P

**Comment by duke8253** [comment]
Well, the default can be changed using a config, discussed with @SolidWallOfCode and we thought this should give the users a bit more control of what timeout to use.

**Comment by zwoop** [comment]
Yeh, but do we really want these to be the same config? Are you positive this is the right way of sharing that setting?

**Comment by SolidWallOfCode** [comment]
I need to double check, but I think this actually changes it back to 10ms, as it was before the event loop changes. I think the global default should be 60ms and let the particular use cases push it down to 10ms.

**Comment by duke8253** [comment]
Yes, this changes it back to how it was before the eventloop changes, I'll make the default to 60ms.

**Comment by zwoop** [comment]
I'm ok with this, I kinda wish though that there was an explanation when and why someone would need / want to change this default from 60 to something else? When do I set it to 100 vs 60 vs 10 ?

**Comment by bryancall** [comment]
@duke8253 Why even have an option to do this?  Shouldn't we just let caller specify what sleep time they want?  Why allow a sleep time of 0, but not a sleep time of 1.

**Comment by duke8253** [comment]
@zwoop The scenario I can think of right now is if someone is experiencing unexpected delays, it might be beneficial for them to tune down this setting and see if it reduce the delay. It will be an indication of what might be wrong and easier for us to pinpoint the cause, and at the same time provide the user a good enough solution to their current delay problem. 

@bryancall This is part of the eventloop changes and this is just the maximum sleep time where the epoll_wait will be waiting. So on a system that's busy this setting won't affect much, but on a system that's not that busy then increasing this config can potentially lower some cpu usage. As for the range from 0 to 1000, I just thought since 0 actually has a meaning for the `epoll_wait`, I'll also add it to the range to support that. I'm also debating whether I should change the range to include -1 since that'll tell `epoll_wait` to not timeout, which might be useful in debugging.

**Comment by SolidWallOfCode** [comment]
Note ithat event loop checks when the next event is scheduled and clips the maximum wait down based on that. In addition because it's `epoll`, actual I/O will break out of the wait and resume activity. The one potentially bad interaction with larger values is with negative events such as the DNS processing if not using a separate DNS thread. I need to think if it's really worth having it configurable beyond debugging and accommodating curmudgeons who say "[you changed it, now it sucks](http://tvtropes.org/pmwiki/pmwiki.php/Main/TheyChangedItNowItSucks)".

**Comment by zwoop** [comment]
@duke8253 Can you add that explanation when one would use this feature, to the docs section?

**Comment by duke8253** [comment]
how about this?

File: doc/admin-guide/files/records.config.en.rst [hunk 1]
```diff
@@ -432,6 +432,14 @@ Thread Variables
    HTTP/2 graceful shutdown for now. Stopping Traffic Server here means sending
    `traffic_server` a signal either by `bin/trafficserver stop` or `kill`.
 
+.. ts:cv:: CONFIG proxy.config.thread.max_heartbeat_mseconds INT 60
```
Comment (SolidWallOfCode) [inline doc/admin-guide/files/records.config.en.rst]: Add `:unit: milliseconds` as per other examples. Pays to be very explicit.

File: doc/admin-guide/files/records.config.en.rst [hunk 2]
```diff
@@ -432,6 +432,14 @@ Thread Variables
...
+
+   Set the maximum heartbeat in milliseconds for threads, ranges from 0 to 1000.
+
+   This controls the maximum wait time of `epoll_wait`. On a system that is not busy,
```
Comment (SolidWallOfCode) [inline doc/admin-guide/files/records.config.en.rst]: Don't say `epoll_wait` - rather something like "maximum amount of time the event loop will wait for I/O activity". Also add some verbiage about this being for expert system tuners, with random fiddling unlikely to be of benefit.

File: doc/admin-guide/files/records.config.en.rst (hunk 1 above)
Comment (duke8253) [reply doc/admin-guide/files/records.config.en.rst]: I added this but github can't seem to find the change.

**Pull request merged**